### PR TITLE
Async Crypto Endeavour

### PR DESCRIPTION
**Pull request by dignifiedquire**
BREAKING CHANGE:
Constructor requires a hash now and there is now Block.create(data, type, cb)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -89,9 +89,17 @@ const Block = require('ipfs-block')
...
+#### `new Block(data, key, [type])`
 
-Creates a new block with raw data `data`. `type` can be either `'protobuf'` or `'ipld'`
+- `data: Buffer|String`
+- `key: Buffer` the multihash of the data
+- `type: String` optional. Defaults to `'protobuf'`
+
+Creates a new block with raw data `data`. `type` can be either `'protobuf'` or `'ipld'`.
+
+#### `Block.create(data, key, [type, ] callback)`
```
Comment (daviddias) [inline README.md]: Let's just go ahead and take from https://github.com/ipfs/js-ipfs-block/pull/5 the fact that .key becomes a function and make it an async function.

File: package.json [hunk 2]
```diff
@@ -38,13 +38,13 @@
   "homepage": "https://github.com/ipfs/js-ipfs-block#readme",
   "devDependencies": {
     "aegir": "^8.0.0",
```
Comment (daviddias) [inline package.json]: update to latest

**Review by daviddias — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (dignifiedquire) [reply README.md]: @diasdavid done, please review the new api before I start propagating

File: package.json (hunk 2 above)
Comment (daviddias) [reply package.json]: 🛎

**Comment by daviddias** [comment]
API looks good, needs:
- [x] a test that tests with a different hash func

File: README.md [hunk 3]
```diff
@@ -25,10 +30,9 @@
   - [Browser: `<script>` Tag](#browser-script-tag)
 - [API](#api)
   - [Block](#block)
-    - [`new Block(data, [type])`](#new-blockdata-type)
+    - [`new Block(data)`](#new-blockdata)
     - [`block.data`](#blockdata)
-    - [`block.key`](#blockkey)
-  - [`block.extension`](#blockextension)
+    - [`block.key([hashFn,] callback)`](#blockkeyhashfn-callback)
```
Comment (daviddias) [inline README.md]: `s/hashFn/hashAlg`

File: README.md [hunk 4]
```diff
@@ -89,26 +93,24 @@ const Block = require('ipfs-block')
...
 #### `block.data`
 
 The raw data of the block. Its format matches whatever was provided in its
 constructor.
 
-#### `block.key`
-
-The [multihash][multihash] of the block's data, as a buffer.
-
+#### `block.key([hashFn,] callback)`
```
Comment (daviddias) [inline README.md]: `s/hashFn/hashAlg`

**Review by daviddias — CHANGES_REQUESTED** [review]
Overall LGTM

**Review by daviddias — APPROVED** [review]

File: src/index.js [hunk 5]
```diff
@@ -14,7 +14,18 @@ function Block (data) {
...
+  data = ensureBuffer(data)
+
+  Object.defineProperty(this, 'data', {
+    get () {
+      return data
+    },
+    set () {
+      throw new Error('Tried to change an immutable block')
+    }
+  })
```
Comment (daviddias) [inline src/index.js:29]: 👍

File: src/index.js [hunk 6]
```diff
@@ -26,7 +37,18 @@ function Block (data) {
       hashFunc = 'sha2-256'
     }
 
-    multihashing(this.data, hashFunc, callback)
+    if (this._cache[hashFunc]) {
+      setImmediate(() => {
```
Comment (Kubuxu) [inline src/index.js]: `setImmediate` isn't supported by FF or Chrome, will it be a problem?

Comment (dignifiedquire) [reply src/index.js]: hmm thanks for pointing this out, need to find a better solution :(

Comment (Kubuxu) [reply src/index.js]: `setImmediate` is also used in 4 different places across the codebase. There is also polyfill for it if it comes to it.

Comment (dignifiedquire) [reply src/index.js]: yeah we were shipping one before, but not anymore. As we already use `async` I'm going to use `async.setImmediate` http://caolan.github.io/async/docs.html#setImmediate

Comment (dignifiedquire) [reply src/index.js]: fixed

**Pull request merged**